### ISSUE TITLE
Transform Performance Slow

### ISSUE THREAD

**Issue by timoxley** [ryoppippi/unplugin-typia#334]
I originally posted this on the typia repo, https://github.com/samchon/typia/issues/1303 but it's probably more appropriate here. 

I have a small vite/vitest application with a handful of tests. The tests take a long time to cold-start and it appears they are getting slowed down by the typia transform: ~15 seconds for ~15 passes, with about 5 files? That seems awfully slow to me, like it must be doing something very inefficient to end up like that?  🤷  What is it doing with all those cpu cycles?

 Is there something that can be tweaked to make it transform faster, e.g. more parallelism, reduce repeated work or something, or is this perhaps a limitation of typescript transform? Any suggestions or additional information I can provide?
 
<p align="center">
 <img src="https://github.com/user-attachments/assets/7ee21b4c-ed9b-43d3-941a-136e7828dee3" alt="~15 seconds for ~15 files" />
 </p>
<p align="center">
  <img src="https://github.com/user-attachments/assets/9a4a3728-83e2-455e-9585-f2c472ea3092" alt="output of vite-plugin-inspect showing typia transform being slow" width="45%">
&nbsp; &nbsp; &nbsp; &nbsp;
  <img src="https://github.com/user-attachments/assets/61882603-a3f1-4ea0-809b-ee5557de6a96" alt="output of vite-plugin-inspect showing files using typia being the slow ones" width="45%">
</p>

AFAICT I'm on the latest version of relevant packages.

```
    "typescript": "^5.6.2",
    "typia": "^6.11.0",
    "vite": "^5.4.4",
    "vite-plugin-inspect": "^0.8.7",
    "vitest": "^2.1.1"
   ```
   
 ----
 
   
 I had a quick poke around the codebase and I wonder if the problem might be related to creating a new ts.host/ts.program for each file (`[id]`), rather than a single program for all the files? 
   
   https://github.com/ryoppippi/unplugin-typia/blob/9877201c8e456f25c2656a9eebd00b53537cfda3/packages/unplugin-typia/src/core/typia.ts#L180
   
  Not sure if that would make any difference, just guessing.
   
  
I also notice that the `unplugin-typia` cache appears very fragile, with any small change it seems to cause a cache miss on everything. I wonder why it's so sensitive? and whether it could use the built-in caches for rollup/vite/tsc etc?  
   
Also, I tried disabling this plugin in vite and using https://github.com/herberttn/vite-plugin-typescript-transform, hoping that it would pick up on the `plugins.transform` in my `tsconfig.json` and do the transforming with the rest of the typescript transpiling, but this didn't seem to work and I'm not sure why.

**Comment by ryoppippi** [comment]
@timoxley OH I missed this issue!
I think the transformation is slow because this transformation depends on tsc. I implemented some extra feature for speed in this plugin like file skipping, caching, and so on. But, the tsc itself is slow so I cannot deal with it.
So, all we can do is improve typia itself!
(Now that oxc can handle dts, I'm thinking it will be faster if rusty-typia can do it, but when will it come out? Or should I just make it ?

**Comment by ryoppippi** [comment]
So as you say, I close this!

**Comment by ryoppippi** [comment]
I'll reopen this!

> I also notice that the unplugin-typia cache appears very fragile, with any small change it seems to cause a cache miss on everything. I wonder why it's so sensitive? and whether it could use the built-in caches for rollup/vite/tsc etc?

I'm curios! If we can use it, this is really great!
In dev mode, if you don't change file, unplugin-typia doesn't process it again, so caching in `vite` works fine.

I'm surprised that it makes slow so much.

**Comment by ryoppippi** [comment]
> caching is fragile.
I agree. That is why I disabled the feature by default..
https://github.com/ryoppippi/unplugin-typia/issues/224

**Comment by ryoppippi** [comment]
Another bad thing is that vite transformation is not parallel, it is sequential.

**Comment by ryoppippi** [comment]
> I had a quick poke around the codebase and I wonder if the problem might be related to creating a new ts.host/ts.program for each file ([id]), rather than a single program for all the files?

Yeah maybe we can fix it. When I first implement this plugin, creating `ts.program` with multiple id didn't work, so I implemented it like this. 
Can we improve it?

**Comment by ryoppippi** [comment]
Oh wait, I found that unplugin-typia gets slower in my project!
Let me think...
![SCR-20240930-sfzv](https://github.com/user-attachments/assets/c651a2ea-b937-4b78-b9ca-9420bec832b7)

**Comment by timoxley** [comment]
hm I tried a few combinations of up/downgrading both the plugin (to 1.0.0) and typia (random versions between 7.x -> 4.0.1) but didn't spot a clear point where the performance degraded significantly. I wonder if it's tsc or some other dependency

**Comment by timoxley** [comment]
> (Now that oxc can handle dts, I'm thinking it will be faster if rusty-typia can do it, but when will it come out? Or should I just make it ?

maybe you should!

**Comment by ryoppippi** [comment]
Hmmm, I'll keep investigating it!
Thank you for your reporting and sponsoring me!!

**Comment by ryoppippi** [comment]
I think typia generates only what you defined in your type definition if you use `assert` function, not `equal`. Does this what you mean?

**Comment by miyaji255** [comment]
If tsc is slow, it could be improved by using `WatchCompilerHost`.

like this:
https://github.com/miyaji255/hono-typia-openapi/blob/main/src/plugin/index.ts

**Comment by ryoppippi** [comment]
@miyaji255 looks interesting 
Do you want to contribute it?

**Comment by ryoppippi** [comment]
@miyaji255 looks interesting 
Do you want to contribute it?

**Comment by bradleat** [comment]
I think we could also save some type by only running the transform if typia is actually imported. Right now it looks like any import from the typia package (like the tags or types) will trigger a transform.

**Comment by zzzz465** [comment]
try using `cache: true` and importing typia to your frontend app entrypoint to make vite (esbuild) optimizes it.
only using `cache: true` causes vite `new dependencies optimized` for `typia/lib/internal` packages.

```ts
import { vitePlugin as remix } from '@remix-run/dev'
import UnpluginTypia from '@ryoppippi/unplugin-typia/vite'
import { flatRoutes } from 'remix-flat-routes'
import { defineConfig } from 'vite'
import OptimizeExclude from 'vite-plugin-optimize-exclude'
import tsconfigPaths from 'vite-tsconfig-paths'

export default defineConfig({
  optimizeDeps: {
    entries: ['./app/**/*.{js,ts,tsx}'],
  },
  plugins: [
    OptimizeExclude(),
    remix({}),
    tsconfigPaths(),
    UnpluginTypia({
      cache: true,
    }),
  ],
})
```

**Comment by ryoppippi** [comment]
yeah actually caching is working.
This is because of `tsc`'s performance actually.
And i don't know how `typia` goes after `typescript-go` come out.

Due to the nature of typia, it cannot help but depend on the speed of the `tsc`. If speed is needed, a new one has to be made.
